Syntex: moving 2022-09-15-preview version from azure-rest-api-specs-pr/RPSaaSMaster to azure-rest-api-specs/main

yurych opened the pull request:
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [x] Other, please clarify - Moving 2022-09-15-preview API version from RPSaaSMaster to main to remove the version from AFEC feature protection
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month. 11/2022
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month. 10/2022
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [x] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [x] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

Commit:
Initial commit of microsoft.syntex to RPSaaSMaster branch (#6063)

* Initial commit of microsoft.syntex to RPSaaSMaster branch
- Dev syntex microsoft.syntex 2021 10 20 preview (#5281)
- Aligned resource type names with the guidelines (plural, camelCase)
- Updated provider name in URIs to lower case and provider type to plural camelCase
- Updated examples to match specification
- Updated DocumentProperties with new required fields, added processDocuments request (#5971)
- Changed suffix from -preview to -privatepreview (#6047)
- Corrected description field (#6059)
- Reverted back to -preview from -privatepreview (#6061)

* Added Syntex to custom-words.txt

* Fixed 'Missing required property: meterId'

* Fixed "the 'openapi-subtype' is unset" error

* Fixing Lint(RPaaS) errors:
- x-ms-long-running-operation
- provisioningState

* More Lint(RPaaS) errors:
- x-ms-long-running-operation-options
- provisioningState directly in the properties bag

* ModelValidation: fixed position of provisioningState

* Fixed OneUnderscoreInOperationId LintDiff errors

* Fixing "Resource provider naming must follow the pascal case" errors.

* Fixing systemData error

* Fixed location of the types.json, which is in common-types

* Fixed location of systemData in the DocumentProcessorResource

* Another re-positioning of systemData

* Started referencing the Resource from common-types

* Changed base resource of DocumentProcessorResource from Resource to TrackedResource

* Corrected update example by adding location property

* correctly placed location property in the update request

* Fixed update LintDiff error

* Fixed Delete 204 LintDiff error

* Fixed XmsIdentifierValidation errors

* Fixed operation response

* Fixed errors related to x-ms-pageable

* Fixed duplicate example Operation IDs

* Fixed x-ms-long-running-operation - related error

* Added description field to ResponseStatus

* Fixed a spelling error

* Fixed azure-asyncoperation - related error

* fixed prettier errors

* Addressed feedback from ARM review board meeting (#6107)

* Removed processDocument operation

* Removed DocumentProcessorProperties and ProvisioningState

* Fixed examples after removing DocumentProperties and ProvisioningState

* Removed x-ms-long-running-operation and 201/202 return codes

* Revert "Removed DocumentProcessorProperties and ProvisioningState"

This reverts commit 2e621471a91c6453a2e9a154604a608dbe059ee4.

* Removed processorName property after reverting DocumentProperties back

* Simplified ProvisioningState to have only 2 states to reflect non-async resource creation status

* Removed depricated status codes from the examples

* Made sure that only tags property of the DocumentProcessorResource can be patched.

* Added "Canceled" terminal state per Sanjai's feedback.

Commit:
Added a new API version to accommodate a new spoTenantAdminUrl property (#7579)

* copied 2021-10-20 to 2022-06-15

* - Added spoTenantAdminUrl property
- Updated examples and the spec to make sure spoAdminTenantUrl is patchable.
- Fixed prettier errors
- copied 2021-10-20 to 2022-06-15
- corrected API version in 2022-06-15 folder
- Restored 2021-10-20 version to its original state
- made the latest API version a default tag in readme.md
- made spoTenantAdminUrl required in the new API (#7586)
- Fixed required property for spoTenantAdminUrl (#7588)

* Corrected PATCH schema and example (#7601)

Commit:
New API version: 2022-09-15-preview (#8815)

* initial copy of 2022-06-15-preview API

* changed API version

* Updated readme.md to incorporate a new version

* Updated resource schema

* removed processDocuments operation, associated types and examples

* Updated examples to match new schema

* Fixed prettier errors

Commit:
Removed API versions that we don't want to move from azure-rest-api-specs-pr/RPSaaSMaster to azure-rest-api-specs/main:
- 2021-10-20-preview
- 2022-06-15-preview
Left only 2022-09-15-preview version.

yurych commented:
Please note:
- the first 3 commits are simply [cherry-picks](https://github.com/Azure/azure-rest-api-specs-pr/commits/RPSaaSMaster/specification/syntex/resource-manager) from azure-rest-api-spec-pr/RPSaaSMaster branch
- the [4th commit](https://github.com/Azure/azure-rest-api-specs/pull/20959/commits/fb1f524c4baba151cf41f4506ab52911f7294b7b) removes API versions that we don't want to move from azure-rest-api-specs-pr/RPSaaSMaster to azure-rest-api-specs/main.

Commit:
Added  "x-ms-parameter-location": "method" to fix XmsParameterLocation LintDiff error.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+        },
+        "spoTenantUrl": {
+          "type": "string",
+          "description": "The URL of an SPO tenant associated with this document processor resource"
```
>SPO [](http://example.com/codeflow?start=40&length=3)

Expand the acronym

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+      "type": "string",
+      "description": "The API version to be used with the HTTP request.",
+      "x-ms-parameter-location": "method"
+    },
```
Reference common parameter definitions in common-types/v3 instead

          "$ref": "../../../../../common-types/resource-management/v3/types.json#/...

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+          "readOnly": true
+        }
+      }
+    }
```
Reference common error response definitions in common-types/v3 instead

          "$ref": "../../../../../common-types/resource-management/v3/types.json#/Definitions/ErrorResponse" etc

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+        }
+      }
+    },
+    "OperationsListResult": {
```
>OperationsListResult [](http://example.com/codeflow?start=5&length=20)

Reference common operations list result in common-types/v3 instead, 

(my best guess for now is

          "$ref": "../../../../../common-types/resource-management/v3/types.json#/Definitions/OperationsListResult"

)

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+      "enum": [
+        "Succeeded",
+        "Failed",
+        "Canceled"
```
Do you have any long-running create or update operations during which you will return additional non-terminal states? What states?

yurych replied:
No, we don't have any other long-running operations.  We discussed this at the ARM review board meeting earlier this year.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+  "info": {
+    "version": "2022-09-15-preview",
+    "title": "Syntex",
+    "description": "Syntex provider"
```
So I notice there is this company called Syntex... is this going to be a potential trademark issue?

TimLovellSmith replied:
Actually I guess its defunct now... I guess that should eliminate the issue but maybe not completely a potential association in people's minds.

TimLovellSmith replied:
The other thing that commonly comes up is that brand names are always changing even at MS. Would it be better to use something generic that won't need rebranding later?

yurych replied:
Syntex is also our brand, we've been working on this provider for a while:
https://www.microsoft.com/en-us/microsoft-365/enterprise/sharepoint-syntex-overview

yurych replied:
The decision to name provider Syntex was made more than a year ago, and we already have private preview customers behind AFEC protection.  We need remove AFEC protection by Ignite, so we have to keep the name as is.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+            "schema": {
+              "$ref": "#/definitions/DocumentProcessorResource"
+            }
+          },
```
PUT needs to return 201 Created responses for initial create, as well as 200 for updates

yurych replied:
It seems that returning 201 makes PUT an asynchronous operation, with the linter validation error:
[RPaaS] An async PUT operation must set long running operation options 'x-ms-long-running-operation-options'

At the ARM review board meeting earlier this year we agreed that this RP does not need asynchronous operations.

TimLovellSmith replied:
See similar thread above, 201 != Async.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+        "systemData": {
+          "$ref": "../../../../../common-types/resource-management/v2/types.json#/definitions/systemData",
+          "readOnly": true
+        },
```
FWIW I hear this is actually optional for now.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+      },
+      "required": [
+        "spoTenantUrl",
+        "spoTenantId"
```
>spoTenantId [](http://example.com/codeflow?start=9&length=11)

Can neither be inferred from the other?

yurych replied:
These two cannot be referred from one another.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+        }
+      }
+    },
+    "DocumentProcessorPatchableProperties": {
```
>DocumentProcessorPatchableProperties [](http://example.com/codeflow?start=5&length=36)

Can possibly be merged with DocumentProcessorProperties.

yurych replied:
They cannot be merged because DocumentProcessorProperties are required and DocumentProcessorPatchableProperties are optional - we need to be able to update one property without affecting the other.  RPaaS validation would not allow updating a single property if DocumentProcessorProperties were used.

TimLovellSmith replied:
I think there might also be a way around that, like only adding the required tags on the DocumentProcessorProperties, but it doesn't really matter.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,521 @@
+      "properties": {
+        "value": {
+          "description": "Collection of available operation details",
+          "uniqueItems": false,
```
>    "uniqueItems": false, [](http://example.com/codeflow?start=6&length=25)

nit, I'm unclear about why they wouldn't be unique

yurych replied:
This is resolved by referring to ""../../../../../common-types/resource-management/v3/types.json#/Definitions/OperationListResult"

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/examples/DocumentProcessor_Create.json:
```diff
@@ -0,0 +1,39 @@
+    }
+  },
+  "responses": {
+    "200": {
```
>200 [](http://example.com/codeflow?start=5&length=3)

Add 201 response example?

yurych replied:
It seems that returning 201 makes PUT an asynchronous operation, and we agreed at the ARM review board meeting earlier this year that this PR will not b asynchronous.

TimLovellSmith replied:
No, returning 201 Created for PUT is just following standard HTTP behavior. Returning 202 makes PUT asynchronous. You may be thinking of returning 201 with nonterminal provisioning status or returning Azure-AsyncOperation headers with 201 which are only needed for async operations. Please review the RPC for further clarification.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/examples/DocumentProcessor_Create.json:
```diff
@@ -0,0 +1,39 @@
+          "additionalProp2": "string2",
+          "additionalProp3": "string3"
+        },
+        "properties": {
```
>properties [](http://example.com/codeflow?start=9&length=10)

Should always have a provisioningState in the response

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/examples/DocumentProcessor_List.json:
```diff
@@ -0,0 +1,44 @@
+              "additionalProp2": "string2",
+              "additionalProp3": "string3"
+            },
+            "properties": {
```
>properties [](http://example.com/codeflow?start=13&length=10)

Should always have a provisioningState in the response

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/examples/DocumentProcessor_List.json:
```diff
@@ -0,0 +1,44 @@
+          {
+            "id": "/subscriptions/cd0d3cc8-cae2-4559-b882-cd707170799b/resourceGroups/mygroup/providers/microsoft.syntex/documentProcessors/myprocessor1",
+            "name": "myprocessor1",
+            "type": "microsoft.syntex/documentprocessors",
```
>syntex [](http://example.com/codeflow?start=31&length=6)

Use consistent capitalization with other resource providers: 'Microsoft.Syntex'

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/examples/DocumentProcessor_Update.json:
```diff
@@ -0,0 +1,38 @@
+    }
+  },
+  "responses": {
+    "200": {
```
>200 [](http://example.com/codeflow?start=5&length=3)

Will your response ever need to run for several seconds or longer? Do you potentially want to return 202 Accepted?

yurych replied:
No, it will not.  This was agreed upon at the ARM board review meeting.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/examples/Operations_List.json:
```diff
@@ -0,0 +1,12 @@
+  "responses": {
+    "200": {
+      "body": {
+        "value": []
```
>value [](http://example.com/codeflow?start=9&length=5)

Should not be an empty example

TimLovellSmith reviewed (changes requested):
:clock1:

Commit:
Tim's feedback:
- Expanded SPO achronym in comments
- Started referencing common-types v3 instead of v2
- Started referencing OperationListResult from common-types v3, removed local OperationsListResult
- Removed systemData from DocumentProcessorResource
- Added provisioningState to Create, Get, and List responses
- Changed capitalization: microsoft.syntex -> Microsoft.Syntex
- Added operation list to the Operations_list.json example

Commit:
Fixed prettier errors

TimLovellSmith commented:
Thanks for the detailed explanation, this is helpful for reviewing.

---
In reply to: [1264133884](https://github.com/Azure/azure-rest-api-specs/pull/20959#issuecomment-1264133884) [](http://example.com/codeflow?ancestors=1264133884,1264133884)

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,493 @@
+            },
+            "spoTenantId": {
+              "type": "string",
+              "description": "The Id (GUID) of an SharePoint Online (SPO) tenant associated with this document processor resource"
```
>Id [](http://example.com/codeflow?start=34&length=2)

nit, ID is preferable to 'Id' in context of documentation

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,493 @@
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          }
```
I may not have been clear in my earlier comemnt but if possible I'd recommend just referencing the latest parameter defnitions from common-types instead i.e.

"$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/SubscriptionIdParameter"

"$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ResourceGroupNameParameter"

"$ref": "../../../../../common-types/resource-management/v3/types.json#/parameters/ApiVersionParameter"

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,493 @@
+      "type": "string",
+      "description": "The API version to be used with the HTTP request.",
+      "x-ms-parameter-location": "method"
+    },
```
Hopefully you don't need to define these, or could just reference definitions in common types, see above.

yurych replied:
Yes, referencing definitions in common-types v3 worked!

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/examples/DocumentProcessor_List.json:
```diff
@@ -0,0 +1,46 @@
+      "body": {
+        "value": [
+          {
+            "id": "/subscriptions/cd0d3cc8-cae2-4559-b882-cd707170799b/resourceGroups/mygroup/providers/microsoft.syntex/documentProcessors/myprocessor1",
```
>microsoft.syntex [](http://example.com/codeflow?start=104&length=16)

I think the consistent capitalization of provider names is preferable in resource ids too - even though they are treated case insensitively.

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,493 @@
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/ErrorResponse"
```
>ErrorResponse [](http://example.com/codeflow?start=37&length=13)

Can potentially refer directly to "$ref": "../../../../../common-types/resource-management/v3/types.json#/definitions/ErrorResponse"

yurych replied:
Thank you, referencing ErrorResponse from common-types v3 worked!

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,493 @@
+    },
+    "DocumentProcessorPatchableProperties": {
+      "type": "object",
+      "description": "Document processor properties that can be patched",
```
>patched [](http://example.com/codeflow?start=64&length=7)

'updated' would be clearer

TimLovellSmith commented on specification/syntex/resource-manager/Microsoft.Syntex/preview/2022-09-15-preview/syntex.json:
```diff
@@ -0,0 +1,493 @@
+          },
+          "x-ms-identifiers": [
+            "name"
+          ],
```
Are document processor resource names globally unique?  If not it would probably be better not to specify 'name' so you get the default of 'id', which would be globally unique

yurych replied:
Thank you, good point, it is not globally unique.

TimLovellSmith reviewed (approved):
🕐

Commit:
Tim's feedback:
- Added 201 response code for PUT request
- Added an example for 201 PUT response

Commit:
Tim's feedback:
- Changed Id -> ID

Commit:
Tim's feedback:
- Started using SubscriptionId, ResourceGroupName and ApiVersion parameters from common-types v3

Commit:
Tim's feedback
- Fixed casing in resource IDs

Commit:
Tim's feedback:
- correciton: patched -> updated

Commit:
Tim's feedback:
- Started referencing ErrorResponse from common-types v3

Commit:
Tim's feedback:
- Removed namd from x-ms-identifiers

yurych commented:
Based on the comment above 
*Please ensure to respond feedbacks from the ARM API reviewer. When you are ready to continue the ARM API review, please remove ARMChangesRequested*
I removed ARMChangesRequested after the PR has been approved, and open-api-workflow-bod added WaitForARMFeedback.  I'm not sure what to do now.
